src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/mapper/HaFlowMapper.java

```diff
@@ -73,4 +81,52 @@ public abstract class HaFlowMapper {
     @Mapping(target = "trackLldpConnectedDevices", ignore = true)
     @Mapping(target = "trackArpConnectedDevices", ignore = true)
     public abstract FlowEndpoint toSubFlowSharedEndpointEncapsulation(HaSubFlow haSubFlow);
+
+    /**
+     * Convert {@link SwitchId} to {@link Switch} object.
+     */
+    public Switch newSwitch(SwitchId switchId) {
+        if (switchId == null) {
+            return null;
+        }
+        return Switch.builder().switchId(switchId).build();
+    }
+
+    public FlowEndpoint map(HaFlow haFlow) {
+        return new FlowEndpoint(haFlow.getSharedSwitchId(), haFlow.getSharedPort(), haFlow.getSharedOuterVlan(),
+                haFlow.getSharedInnerVlan());
+    }
+
+    /**
+     * Converts {@link HaFlowRequest} to a few {@link RequestedFlow}.
+     */
+    public Collection<RequestedFlow> toRequestedFlows(HaFlowRequest request) {
+        return request.getSubFlows().stream()
+                .map(subFlow -> RequestedFlow.builder()
+                        .flowId(subFlow.getFlowId())
+                        .haFlowId(request.getHaFlowId())
+                        .srcSwitch(request.getSharedEndpoint().getSwitchId())
+                        .srcPort(request.getSharedEndpoint().getPortNumber())
+                        .srcVlan(request.getSharedEndpoint().getOuterVlanId())
+                        .srcInnerVlan(request.getSharedEndpoint().getInnerVlanId())
+                        .destSwitch(subFlow.getEndpoint().getSwitchId())
+                        .destPort(subFlow.getEndpoint().getPortNumber())
+                        .destVlan(subFlow.getEndpoint().getOuterVlanId())
+                        .destInnerVlan(subFlow.getEndpoint().getInnerVlanId())
+                        .detectConnectedDevices(new DetectConnectedDevices())
+                        .description(subFlow.getDescription())
+                        .flowEncapsulationType(request.getEncapsulationType())
+                        .bandwidth(request.getMaximumBandwidth())
+                        .ignoreBandwidth(request.isIgnoreBandwidth())
+                        .strictBandwidth(request.isStrictBandwidth())
+                        .pinned(request.isPinned())
+                        .priority(request.getPriority())
+                        .maxLatency(request.getMaxLatency())
+                        .maxLatencyTier2(request.getMaxLatencyTier2())
+                        .periodicPings(request.isPeriodicPings())
+                        .pathComputationStrategy(request.getPathComputationStrategy())
+                        .allocateProtectedPath(request.isAllocateProtectedPath())
+                        .build())
+                .collect(Collectors.toCollection(() -> new TreeSet<>(Comparator.comparing(RequestedFlow::getFlowId))));
```

Comment:
It is somehow important that they are in sorted order? It doesn't seem to be covered in contract, type, or a unit test.

Reply:
Ah, it's like `.distinct()` using some field. Nice trick!
Anyway, could be useful to add this explanation to the javadoc,

Reply:
great, thanks. Just a typo there: ensure